Example/ViewController.swift

```diff
@@ -48,7 +54,237 @@ class ViewController: UIViewController {
             startButton.isEnabled = true
             mapView?.show(routes)
             mapView?.showWaypoints(on: currentRoute)
+            if let style = mapView?.style {
+                // show route Duration and Toll annotations
+                updateRouteAnnotations(routes, style: style)
+            }
+        }
+    }
+
+    fileprivate let dateComponentsFormatter = DateComponentsFormatter()
+
+    private func updateAnnotationSymbolImages(_ style: MGLStyle) {
+        let capInsetHeight = CGFloat(22)
+        let capInsetWidth = CGFloat(11)
+        let capInsets = UIEdgeInsets(top: capInsetHeight, left: capInsetWidth, bottom: capInsetHeight, right: capInsetWidth)
+        if let image = UIImage(named: "RouteInfoAnnotationLeftHanded") {
+            let regularRouteImage = image.tint(UIColor.white).resizableImage(withCapInsets: capInsets, resizingMode: .stretch)
+            style.setImage(regularRouteImage, forName: "RouteInfoAnnotationLeftHanded")
+
+            let selectedRouteImage = image.tint(#colorLiteral(red: 0.337254902, green: 0.6588235294, blue: 0.9843137255, alpha: 1)).resizableImage(withCapInsets: capInsets, resizingMode: .stretch)
+            style.setImage(selectedRouteImage, forName: "RouteInfoAnnotationLeftHanded-Selected")
+        }
+
+        if let image = UIImage(named: "RouteInfoAnnotationRightHanded") {
+            let regularRouteImage = image.tint(UIColor.white).resizableImage(withCapInsets: capInsets, resizingMode: .stretch)
+            style.setImage(regularRouteImage, forName: "RouteInfoAnnotationRightHanded")
+
+            let selectedRouteImage = image.tint(#colorLiteral(red: 0.337254902, green: 0.6588235294, blue: 0.9843137255, alpha: 1)).resizableImage(withCapInsets: capInsets, resizingMode: .stretch)
+            style.setImage(selectedRouteImage, forName: "RouteInfoAnnotationRightHanded-Selected")
+        }
+    }
+
+    private func updateRouteAnnotations(_ routes: [Route]?, style: MGLStyle) {
```

Comment:
This method is making pretty heavy computation each time. I'd probably consider outlining only specific actions when we should update delay view location (e.g. it doesn't make sense to show delay view on low zoom levels, since route is not really visible anyway). We can also move certain calculations to background thread.

Reply:
@1ec5 @MaximAlien - This brings up a question I have. Can we perform map style updates from a background thread?

Reply:
Actually, I don't think this is actually that heavy in terms of performance. I profiled several recomputations of the route annotation coordinates by repeatedly scrolling the map view (which causes a re-layout) and the time spent in this routine is only about 1% of the CPU time. It also will only happen when the map is moved or otherwise redrawn.

It's also much, much smaller than the delay in retrieving the new route from the Directions API so causes only imperceptible delays in the case when a new route is being fetched. 

I'd argue it's better just to leave it as is than to try and add in some concurrency.

Hopefully @1ec5 can weigh in on this.

Reply:
It may be worth moving the calculations to a background thread just in case, but most likely the reason it hasn’t come up as a performance bottleneck yet is that it only runs once per camera change and not continuously (which is fine).